[Aiken Wong Xiheng] iP

aikenwx opened the pull request:
# DukeBot
> “Your mind is for having ideas, not holding them.” – David Allen ([Source](https://dansilvestre.com/productivity-quotes))

DukeBot frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use

All you need to do is,

1. download it from [here](https://github.com/aikenwx/ip/releases/tag/v0.2).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE!**

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events 
- [x] Unit testing 
- [x] GUI implementation
- [x] Sort your tasks chronologically
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

aikenwx committed:
Level-1

aikenwx committed:
Level-2

aikenwx committed:
Level-3

aikenwx committed:
Level-4

aikenwx committed:
A-TextUiTesting

aikenwx committed:
Level-5

aikenwx committed:
Level-6

aikenwx committed:
A-Enums

aikenwx committed:
Update initialise date/timeline keywords in taskFactory method

'/' keyword does not meet the '/at' and '/by' keyword requirements set
for initialising an event and a deadline repectively.

Let's,
* update taskFactory method to accept '/at' and '/by' keywords when
  initializing Event and Deadline classes respectively.
* update input.txt and EXPECTED.TXT to reflect new changes in tests too.

aikenwx committed:
Add save task list to hard disk feature

Task list is not save to hard disk and is reset each time the Duke stops
running.

Saving the task list to the hard disk allows for future retrieval each
time Duke reloads.

Let's create a FileManager class to handle saving and retrieval of task
lists.

aikenwx committed:
Add parsing and interpretation for dates and time

Date and time stored as per input string, which does not allow Duke to
use such data in more meaningful ways.

By restricting format of input dates and time and then parsing the data
into java.time.LocalDateTime, we can perform useful data manipulation
like finding events and deadlines by date.

Let's
* Add date fields for Event and Deadline
* Add parsing for input dates and time into java.time.LocalDateTime
* Add find by date command
* Update tests to test date formatting and parsing and find command

aikenwx committed:
Merge branch 'branch-level-7'

aikenwx committed:
Merge branch 'branch-level-8'

aikenwx committed:
Extract Duke into Parser, Ui and command classes:

Duke contains many methods such as addTask and markAsDone that can be
similarly grouped. Furthermore many UI and parsing related methods
exists throughout the programme that can be similarly grouped to avoid
code duplication.

Extracting such similar behaviour into seperate classes create distinct
roles for each class and prevents Duke from becoming too large with
variious categories of methods with different behaviours.

Let's
* Pull command operations into various command classes with Command as
  the parent abstract class.
* Create a Ui class to handle all UI related operations.
* Crete a Parser class to handle all parsing related operations.
* Rename FileManager class to Storage class to better represent the use
  case for the class.

aikenwx committed:
Create duke, command and task packages

aikenwx committed:
Add jUnit tests for Parser and Storage

aikenwx committed:
Create JAR file for Duke

aikenwx committed:
Add JavaDoc comments to Duke

aikenwx committed:
Add find by keyword functionality

aikenwx committed:
Merge branch 'branch-A-JavaDoc' into mergeToThis

aikenwx committed:
Check Coding Standard

aikenwx committed:
Merge branch 'branch-A-CodingStandard' into mergeToThis

aikenwx committed:
Merge branch 'branch-Level-9' into mergeToThis

aikenwx committed:
Update expected test results

aikenwx committed:
Correct violatations to Java coding standard.

aikenwx committed:
Merge branch 'add-gradle-support' of https://github.com/aikenwx/ip

aikenwx committed:
Add gradle support to Duke

aikenwx committed:
Add CheckStyle support to Duke

aikenwx committed:
Correct more coding standard violations

aikenwx committed:
Add GUI to Duke

aikenwx committed:
Add visual improvements to GUI

aikenwx committed:
Merge branch 'branch-A-Gradle'

aikenwx committed:
Merge branch 'branch-A-CheckStyle'

aikenwx committed:
Host Duke on GUI implementation

aikenwx committed:
Fix coding violations

aikenwx committed:
Add more JavaDoc documentation

aikenwx committed:
Merge branch 'branch-Level-10'

aikenwx committed:
Refactor code to improve readability and reduce code duplication

nwhloo99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /**
+     * This method runs the app.
```
Should the header comment start with `Runs...`?

nwhloo99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,243 @@
+     * @throws DukeException
+     */
+    public static LocalDateTime parseDateAndTime(String dateInput, String timeInput) throws DukeException {
+
```
Should the white lines here be removed?

nwhloo99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,140 @@
+package duke;
+
```
I noticed that there's always 2 blank lines between the package and the imports, perhaps you can set your IDE settings to make it 1, so you do not have to worry about this in the future.

nwhloo99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,204 @@
+     * Produces a string enumerating the tasks of the given task list.
+     *
+     * @param tasks The given tasks list to enumerate.
+     * @return
```
DId you forget to add the return section here?

nwhloo99 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+import duke.Ui;
+import duke.task.Task;
+import duke.task.TaskList;
+
```
Should there be javadocs for each of the command classses?

nwhloo99 reviewed (commented):
I like that your code is quite clean and follows the coding standards quite well. I think you just need to be careful of small careless mistakes, like adding too many new lines or forgetting a small part of your JavaDocs. Otherwise, your code is quite well done! 👍

aikenwx committed:
Combine GraphicalUserInterface and Ui into one

bentanjunrong commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,73 @@
+        this.setAlignment(Pos.TOP_RIGHT);
+        this.getChildren().addAll(text, displayPicture);
+    }
+
```
Minor (Coding Standard): Inconsistent spacing between lines. I believe it should be one line of space between functions 😬

bentanjunrong commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,73 @@
+ * @author Aiken Wong
+ */
+public class DialogBox extends HBox {
+
```
Minor (Coding Standard): No need for line space here 😬

bentanjunrong commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,73 @@
+        db.flip();
+        return db;
+    }
+
```
Minor (Coding Standard): No need for line breaks here!

bentanjunrong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        } catch (DukeException e) {
+            System.out.println(e.getMessage());
+        }
+
```
Minor (Coding Standard): No need for line breaks here!

bentanjunrong commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @author Aiken Wong
+ */
+public class DukeException extends RuntimeException {
```
Personally don't think RuntimeException should be thrown here? Based on the project description, when an error is encountered, the user should be allowed to continue executing the program. You could instead inherit from the generic Exception class instead

bentanjunrong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,243 @@
+
+    private static String errorMessage = "Wrong format Sir/Mdm. Dates and times must be given as only a date: DATE\n"
+        + "or as date and time: DATE TIME\n"
+        + "Accepted formats for DATE: YYYY-MM-DD, DD/MM/YYYY\n"
```
Nothing wrong with this, but there's no need to end your substrings with "\n". You could have combined the substrings together to make it more compact!

bentanjunrong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,140 @@
+            Task currentTask = tasks.get(i);
+            if (currentTask instanceof Event) {
+                Event event = (Event) currentTask;
+                text += String.format("| E | %s | %s | %s | %s | %s\n", event.getIsDone() ? "X" : " ",
```
I feel like it might be more aligned to OOP principles to have this String parsing done within the specific Task classes, and just calling this method here e.g. public String toFileString() { return String.format( ...); }

bentanjunrong commented:
Overall I think the code looks good! OOP principles were followed, and it's written in a logical and clean way. Well done!

aikenwx committed:
Add sort command

Tasks in tasklist are unsorted and arranged in order of addition by the
user.

Giving users the ability to sort their tasks in chonological order
will enable them see view their task in the order of most urgent to
least urgent.

Chronological ordering is prefered over lexicographic ordering since
users will be completing tasks in order of the next upcoming task
anyway.

Let's create a new sort command that enables users to sort their tasks
in chronological order.

aikenwx committed:
Add assertions to Storage and Event classes

aikenwx committed:
Add command: add exception for unknown task type

The getTaskType identifies the added task as a Deadline if it is not
instance of an Event or Todo.

However, it does not handle the case if the task is also not a Deadline.

Adding an exception at the point of identifying an unknown task type
will catch the unexpected behaviour early, allowing for better debugging
in case of future regressions.

Let's add and throw an exception upon identification of an unknown
TaskType in the getTaskType method of AddCommand.

aikenwx committed:
Refactor Storage class and all its dependencies

aikenwx committed:
Refactor Parser class and all its dependencies

aikenwx committed:
Move building of storage string from Storage to TaskList

The buildStorageString method in Storage implicitly understands the
structure of TaskList and Task to be able to build the storage string.

Following OOP principles, TaskList should be able to generate the
storage string on its own and not expose its implementation.

Let's refactor Storage and TaskList and move buildStorageString method
into TaskList as getStorageString.

aikenwx committed:
Refactor DukeException to extend Exception rather than RuntimeException

aikenwx committed:
Merge pull request #2 from aikenwx/branch-A-Assertions

Add assertions to Storage and Event classes

aikenwx committed:
Merge branch 'master' into branch-A-CodeQuality

aikenwx committed:
Merge pull request #4 from aikenwx/branch-A-CodeQuality

 Improve code quality

aikenwx committed:
Merge branch 'master' into branch-C-Sort

aikenwx committed:
Merge pull request #3 from aikenwx/branch-C-Sort

Add sort command

aikenwx committed:
Update user and bot GUI display photos

aikenwx committed:
Add error reply dialogue box, update background photo

aikenwx committed:
fix background image and dukeReply user dialogue text position

aikenwx committed:
Refactor Ui and DialogBox classes

aikenwx committed:
Update DialogBox font and app name

aikenwx committed:
Add styling for input text field and send button

aikenwx committed:
Merge branch 'branch-A-BetterGui'

aikenwx committed:
Add Ui.png into docs folder

aikenwx committed:
Update user guide in README.md

aikenwx committed:
Add images and summary to user guide

aikenwx committed:
Update image alignment in README.md

aikenwx committed:
Set theme jekyll-theme-slate

aikenwx committed:
Correct punctuation in README.md

aikenwx committed:
Fix incorrect resource PATH in Ui class
